crypto/mem_sec.c

```diff
@@ -433,8 +438,19 @@ static int sh_init(size_t size, int minsize)
     if (mprotect(sh.map_result + aligned, pgsize, PROT_NONE) < 0)
         ret = 2;
 
+#if defined(OPENSSL_SYS_LINUX) && defined(MLOCK_ONFAULT) && defined(SYS_mlock2)
+    if (syscall(SYS_mlock2, sh.arena, sh.arena_size, MLOCK_ONFAULT) < 0) {
+        if (errno == ENOSYS) {
+            if (mlock(sh.arena, sh.arena_size) < 0)
+                ret = 2;
+        } else {
```

Comment:
Based on my (limited) testing, the secure memory region when used with mlock() allocates it all and wires it , such that RSS = kbytes as viewed in pmap(1). When mlock2(MLOCK_ONFAULT) is used, RSS < kbytes. This demonstrates to me that MLOCK_ONFAULT had an effect. The man page for mlock indicates: 

> mlock() locks pages in the address range starting at addr and continuing for len bytes. All  pages that contain a part of the specified address range are guaranteed to be resident in RAM when the call returns successfully; the pages are guaranteed to stay in RAM until later unlocked.

This indicates to me (and demonstrated via pmap(1)) that mlock does the actual wiring of RAM and pages do not need to be touched.

That being said, the only reason I can think of mlock2() failing and mlock() succeeding is when mlock2(MLOCK_ONFAULT) does not exist. Conversely, I can see mlock() failing due to a lack of physical memory being available, but mlock2(MLOCK_ONFAULT) succeeding.